### PR TITLE
PERF: releasing the GIL, #8882

### LINKED ISSUE #8882: Release the GIL in Cython code
Pandas could release the GIL in parts of the Cython code.  This would enable the parallel use of multiple cores with threading.

From off-line conversation with @jreback @cpcloud 

http://lbolla.info/blog/2013/12/23/python-threads-cython-gil

### PR DESCRIPTION
**Pull request by jreback**
closes #8882
closes #10139 

This is now implemented for all groupbys with the factorization part as well (which is actually a big part fo the time)

```
In [2]: N = 1000000

In [3]: ngroups = 1000

In [4]: np.random.seed(1234)

In [5]: df = DataFrame({'key' : np.random.randint(0,ngroups,size=N),
   ...:                 'data' : np.random.randn(N) })
```

```
def f():
    df.groupby('key')['data'].sum()

# run consecutivily
def g2():
    for i in range(2):
        f()
def g4():
    for i in range(4):
        f()
def g8():
    for i in range(8):
        f()

# run in parallel
@test_parallel(num_threads=2)
def pg2():
    f()

@test_parallel(num_threads=4)
def pg4():
    f()

@test_parallel(num_threads=8)
def pg8():
    f()
```

So seeing a nice curve as far as addtl cores are used (as compared to some of my previous posts).

```
In [19]: df
Out[19]: 
   after  before   speedup
2   27.3    52.1  1.908425
4   44.8   105.0  2.343750
8   83.2   213.0  2.560096
```

### CODE REVIEW

**Comment by mrocklin** [comment]
What was the deal with `NA` and `nan`.  These were Python objects?

**Comment by jreback** [comment]
apparently doing something

```
cdef ndarray[double_t] foo

foo = np.empty(10,dtype='float64')

foo[0] = np.nan
```

actually assigns a python object (and does inference and such), so that you MUST have the gil,
but casting to a double seems to work though (as its now c-level)

**Comment by ahojnnes** [comment]
You can reuse NPY_NAN cross-platform, e.g., https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/_texture.pyx#L11

**Comment by jreback** [comment]
@ahojnnes thxs, I did change it.

**Comment by jreback** [comment]
I updated this. works for all groupbys now. So it gives a general 5-10% speedup on most operations (even single threaded), as there is slightly less generated code (in ``nogil`` it pretty much generates straight c code).

so we still need to hold the gil for the ``.resize`` operation, but its generally quite limited when this happens (e.g. resizing is a function of the number of uniques for example), but the cost of this is pretty low.

There IS a soln where I can not use numpy resize at ALL. But its quite a bit more c-code and IMHO not really worth it.

**Comment by mrocklin** [comment]
Cool.  I'll try to run a few different groupby operations with dask.dataframe sometime early next week and report back.  I'm curious what were the operations that needed to be released?  Was there a core set of functionality or was it a large volume of trivial work?

**Comment by jreback** [comment]
TL;DR.

So lots of 'boilerplate' changes in ``generated.pyx``, which is basically all of the algorithms for groupby ops themselves (e.g. sum,mean). These were straightforward insertions of ``with nogil``, just needed to avoid typing the ``object`` ones (which for the most part was easy; though a bit tricky on some non-groupby routines, e.g. see #10213).

However, the big speeds occur once I fixed the ``hashtable.pyx`` factorizer (e.g. the ``get_labels`` routine). The trivial soln actually produces a very odd perf issue (see [here](https://groups.google.com/forum/#!topic/cython-users/jtZ0FOGX2vU)), which is a bug / documentation issue.

In effect, you are using a ``nogil`` function that happens to have a ``with gil`` block WITHIN it, then the function is way slower than if you inline the exact same code.

I 'fixed' this issue by adding some additional c-structures to hold data (basically I changed the low-level implementation); these can be easily passed around; downside is that are a little 'messy' in the code department.

note to @shoyer I didn't actually bypass the realloc of memory (e.g. ``.resize``) of numpy arrays. though I do have a soln to that (basically manage a memory block myself, then turning that into a numpy array at the very end (in ``to_array``), easy enough, but was running into some compile issues, so dropped it in the end).

**Comment by mrocklin** [comment]
I think that this is awesome and am afraid of it going stale.  Is there something blocking it from being merged into master?

**Comment by shoyer** [comment]
@mrocklin We were waiting to get the 0.16.2 release out first, which happened last Friday. At this point I don't think there are any blockers.

**Comment by jreback** [comment]
nothing blocking. I am going to update and see if I can hit some of #10213.

I have to do a bit of odd code-generation because for example you want to simultaneously support ``nogil`` (for numeric) and ``gil`` for ``object``, but its a context manager (and NOT a function, which has some odd per characteristics). IOW, you have to worry about formatting, a bit annoying.

**Comment by jreback** [comment]
actually I do need to add a doc-note to the whatsnew (and maybe a small mention in the enhancingperf section as well).

**Comment by mrocklin** [comment]
Hrm, yes, unfortunate that you can't manipulate the lock directly

**Comment by jreback** [comment]
yep, would be nice to put in something along the lines of an ``#IFDEF``, but in reality the block is actually doing 'work', so maybe not easy/possible.

**Comment by mrocklin** [comment]
Thought I'd caress this PR with a gentle ping.  

*ping*

**Comment by jreback** [comment]
is that euphemism for harassment?

**Comment by mrocklin** [comment]
Kind and loving harassment, yes.

**Comment by jreback** [comment]
any commentary @jorisvandenbossche @shoyer @cpcloud @TomAugspurger 

going to have a followup for more GIL stuff. Its actually tricky making sure its doing (and you are measure exactly the right stuff)

File: doc/source/whatsnew/v0.17.0.txt [hunk 1]
```diff
@@ -56,8 +57,32 @@ Deprecations
...
+.. _dask: https://dask.readthedocs.org/en/latest/
+
+.. _whatsnew_0170.gil:
+
+Releasing the GIL
+~~~~~~~~~~~~~~~~~
+
+We are releasing the global-interpreter-lock (GIL) on some cython operations.
+This will allow other threads to run simultaneously during computation, potentially allowing performance improvements
+from multi-threading. Notably ``groupby`` and some indexing operations are a benefit to this. (:issue:`8882`)
```
Comment (cpcloud) [inline doc/source/whatsnew/v0.17.0.txt]: i'd say "benefit from this"

File: doc/source/whatsnew/v0.17.0.txt [hunk 2]
```diff
@@ -56,8 +57,32 @@ Deprecations
...
+as well as the ``.sum()`` operation.
+
+.. code-block:: python
+
+   N = 1e6
+   df = DataFrame({'key' : np.random.randint(0,ngroups,size=N),
+                   'data' : np.random.randn(N) })
+   df.groupby('key')['data'].sum()
+
+The releasing of the GIL could benefit an application that uses threads for user interactions (e.g. ``QT``), or performaning multi-threaded computations. A nice example of a library that can handle these types of computation-in-parallel is the dask_ library.
```
Comment (cpcloud) [inline doc/source/whatsnew/v0.17.0.txt]: Releasing the GIL instead of "the releasing of the GIL"

File: doc/source/whatsnew/v0.17.0.txt [hunk 3]
```diff
@@ -56,8 +57,32 @@ Deprecations
...
+.. _whatsnew_0170.gil:
+
+Releasing the GIL
+~~~~~~~~~~~~~~~~~
+
+We are releasing the global-interpreter-lock (GIL) on some cython operations.
+This will allow other threads to run simultaneously during computation, potentially allowing performance improvements
+from multi-threading. Notably ``groupby`` and some indexing operations are a benefit to this. (:issue:`8882`)
+
+For example the groupby expression in the following code will have the GIL released during the groups computation, e.g. ``df.groupby('key')``
```
Comment (cpcloud) [inline doc/source/whatsnew/v0.17.0.txt]: maybe say "during the factorization step" instead of "during the groups computation", not totally clear what the "groups" computation is

File: pandas/hashtable.pyx [hunk 4]
```diff
@@ -80,72 +61,117 @@ cdef class ObjectVector:
...
+        self.ao.resize(self.n)
+        self.m = self.n
+        return self.ao
+
+
+ctypedef struct Int64VectorData:
+    int64_t *data
+    size_t n, m
+
+cdef uint8_t Int64VectorData_needs_resize(Int64VectorData *data) nogil:
```
Comment (cpcloud) [inline pandas/hashtable.pyx]: I think you can used fused types to reduce the duplicity here.

```cython
ctypedef fused vector_data:
    Int64VectorData
    Float64VectorData

ctypedef fused sixty_four_bit_scalar:
    int64_t
    float64_t

cdef bint needs_resize(vector_data *data) nogil:
    return data.n == data.m

cdef void append_data(vector_data *data, sixty_four_bit_data x):
    data.data[data.n] = x
    data.n += 1
```

File: pandas/hashtable.pyx [hunk 5]
```diff
@@ -399,137 +324,166 @@ cdef class Int64HashTable: #(HashTable):
         else:
             raise KeyError(key)
 
-    def map(self, ndarray[int64_t] keys, ndarray[int64_t] values):
+    @cython.boundscheck(False)
```
Comment (cpcloud) [inline pandas/hashtable.pyx]: do these also need `cython.wraparound(False)`?

File: pandas/hashtable.pyx [hunk 6]
```diff
@@ -740,7 +719,7 @@ cdef class PyObjectHashTable(HashTable):
             int ret = 0
             object val
             khiter_t k
-            ndarray[int64_t] locs = np.empty(n, dtype=np.int64)
+            int64_t[:] locs = np.empty(n, dtype=np.int64)
```
Comment (cpcloud) [inline pandas/hashtable.pyx]: Was this change necessary or just for consistency with the rest of the HashTable implementations?

File: pandas/hashtable.pyx (hunk 5 above)
Comment (jreback) [inline pandas/hashtable.pyx]: yes when you release the gil it gives a warning if you don't do this (as you are de-facto turning it off in c-land anyhow, I think it wants to make it explict)

File: pandas/hashtable.pyx (hunk 6 above)
Comment (jreback) [inline pandas/hashtable.pyx]: just consistency

File: pandas/hashtable.pyx [hunk 7]
```diff
@@ -904,28 +864,34 @@ cdef class Int64Factorizer:
...
+                table.vals[k] = 1
 
 
-cpdef value_count_int64(ndarray[int64_t] values):
+@cython.boundscheck(False)
+cpdef value_count_int64(int64_t[:] values):
     cdef:
         Py_ssize_t i
         kh_int64_t *table
+        ndarray[int64_t] result_keys, result_counts
```
Comment (jreback) [inline pandas/hashtable.pyx]: I think I tried to change this before and ran into an issue..trying again.

File: pandas/hashtable.pyx [hunk 8]
```diff
@@ -80,72 +61,120 @@ cdef class ObjectVector:
...
+
+cdef bint needs_resize(vector_data *data) nogil:
+    return data.n == data.m
+
+cdef void append_data(vector_data *data, sixty_four_bit_scalar x) nogil:
+
+    # compile time specilization of the fused types
+    # as the cross-product is generated, but we cannot assign float->int
+    # the types that don't pass are pruned
+    if (vector_data is Int64VectorData and sixty_four_bit_scalar is int64_t) or (
```
Comment (jreback) [inline pandas/hashtable.pyx]: @cpcloud I got this to work after puzzling on what cython does with multiple fused types. It creates the cross-product, but you generally need only certain specilizations; it does this beaut as a compile-time prune, so no perf hit.

**Comment by sinhrks** [comment]
Great job!!

**Comment by mrocklin** [comment]
Woohoo!

*Pulls from master, checks benchmarks.*

**Comment by jorisvandenbossche** [comment]
@mrocklin Can we already look forward to a new awesome blogpost? :-)

**Comment by mrocklin** [comment]
Best I have to offer at the moment are the slowly growing [`dask.dataframe` docs](http://dask.pydata.org/en/latest/dataframe.html)

**Comment by jorisvandenbossche** [comment]
I should definitely check it out once I find some time!

**Comment by scari** [comment]
Wow! Great job! 👍🏻
